### ISSUE TITLE
Re-review encryption and/or concurrency-safety in the copy pipeline

### ISSUE THREAD

**Issue by mtrmac** [containers/image#999]
https://github.com/containers/image/blob/7810876ebe28eb6ed405dbe825159b22781cf813/copy/copy.go#L1139-L1140 looks fairly clearly wrong, writing to the per-copy shared data inside a per-image code, executed concurrently inside the per-layer copy, even.

So, re-review the data flow for encryption, and perhaps audit and better document the flow/concurrency-safety overall.

**Comment by lumjjb** [comment]
Would setting this assignment in
```
https://github.com/containers/image/blob/7810876ebe28eb6ed405dbe825159b22781cf813/copy/copy.go#L602
```
solve the issue here? It looks the `ic.oci*Config` fields are there for some reason no longer needed. Let me write up a PR to fix this.

**Comment by mtrmac** [comment]
My initial guess is that the fields should be initialized in `copier` when creating it, and not present in `imageCopier` at all.

This issue also serves as a more general reminder to check what else I should have paid attention to — and to document how it should work for future contributors.

**Comment by mtrmac** [comment]
This was mostly fixed in #1047 ; and later https://github.com/containers/image/pull/2048 removed the private references to encryption state entirely, sharing the original pointers directly.

Reviewing the concurrent uses of that data, the only potentially risky place I can find is https://github.com/containers/ocicrypt/blob/ffc163c18e970386fcd1b35acf2d350c8665d080/keywrap/pkcs11/keywrapper_pkcs11.go#L44 , where the `append` could potentially be called concurrently, and if the original slice had extra capacity, that would mean that concurrent encryptions may write to the same extra fields of that slice.

That seems to be possible — but that needs to be fixed in `ocicrypt`: https://github.com/containers/ocicrypt/issues/92 .
